tests/kernel_bundle/sycl_compile.h

```diff
@@ -0,0 +1,165 @@
+    return;
+  }
+
+  auto input_kb =
+      sycl::get_kernel_bundle<sycl::bundle_state::input>(ctx, {dev}, {k_id});
+  auto input_ids = input_kb.get_kernel_ids();
+
+  auto obj_kb = compile_bundle<Overload>(input_kb, {dev});
+  kernel_bundle::define_kernel<DescriptorT, sycl::bundle_state::input>(queue);
+
+  auto restrictions{
+      kernel_bundle::get_restrictions<DescriptorT,
+                                      sycl::bundle_state::input>()};
+  bool dev_is_compat = restrictions.is_compatible(dev);
+
+  if (obj_kb.has_kernel(k_id) != dev_is_compat) {
+    FAIL(log, "Device does not support kernel " + kName +
+                  " but compiled bundle"
+                  " contains it");
+  }
+
+  // Check that result object bundle has the same kernels as input bundle
+  bool same_kernels_in_obj_bundle = true;
+  for (const auto &in_id : input_ids) {
+    same_kernels_in_obj_bundle &= obj_kb.has_kernel(in_id);
+  }
+
+  if (!same_kernels_in_obj_bundle) {
+    FAIL(log,
+         "Result bundle does not contain all kernels from input bundle "
+         "(kernel: " +
+             kName + ")");
+  }
+
+  // Check that input and result kernel bundles have the same context
+  if (input_kb.get_context() != obj_kb.get_context()) {
+    FAIL(log,
+         "Input bundle and result bundle have different contexts "
+         "(kernel: " +
+             kName + ")");
+  }
+}
+
+/** @brief Check that sycl::compile result bundle has the same associated
+ *         devices as presented in input device vector
+ *  @tparam Overload Selected overload of sycl::compile
+ */
+template <CompileOverload Overload>
+static void check_associated_devices(util::logger &log) {
+  auto queue = util::get_cts_object::queue();
+  auto ctx = queue.get_context();
+  std::vector<sycl::device> devices{queue.get_device()};
+  // Force duplicates
+  devices.push_back(devices[0]);
+
+  // Test can be skipped if all devices do not support online compilation
+  // since this is not sycl::compile's fault
+  if (!sycl::has_kernel_bundle<sycl::bundle_state::input>(ctx)) {
+    log.skip("No kernel bundle with input state for test (skipped).");
+    return;
+  }
+
+  auto input_kb =
+      sycl::get_kernel_bundle<sycl::bundle_state::input>(ctx, devices);
+  auto obj_kb = compile_bundle<Overload>(input_kb, devices);
+
+  // Check that result kernel bundle contains all devices from passed vector
+  bool same_devs = true;
+  auto kb_devs = obj_kb.get_devices();
+  for (const auto &dev : devices) {
+    same_devs &=
+        (std::find(kb_devs.begin(), kb_devs.end(), dev) != kb_devs.end());
+  }
+  if (!same_devs) {
+    FAIL(log,
+         "Set of associated to obj_kb devices is not equal to list of "
+         "devices passed.");
+  }
+
+  // Check that result kernel bundle does not have duplicates
+  bool unique_devs = true;
+  for (const auto &dev : kb_devs) {
+    unique_devs &= (std::count(kb_devs.begin(), kb_devs.end(), dev) == 1);
+  }
+  if (!unique_devs) {
```

Comment:
Use an algorithm

Reply:
Yes, this one is subtler.
Right now you have an O(n^2) implementation.
By looking at https://en.cppreference.com/w/cpp/algorithm, what about:
```c++
std::sort(kb_devs.begin(), kb_devs.end());
if (std::adjacent_find(kb_devs.begin(), kb_devs.end()) != kb_devs.end()) {
```
in O(n log n)?
Perhaps we could have the utility function to return always a sorted list at the first place?
@illuhad interesting to see all this code requiring comparison of devices. :-)

Reply:
This is either a bug in the implementation or a bug in the SYCL spec.
It should be possible to have a `std::set` or even a `std::unsorted_set` of `sycl::device`.

Reply:
> It should be possible to have a `std::set` or even a `std::unsorted_set` of `sycl::device`.
@keryell We can use `std::unsorted_set<sycl::device>` without some errors, and we can use the `std::set<sycl::device>` with a comparing function.

But in this case we can't use `std::sort` for sorting by `std::hash` because of the fact that result of the `std::hash` is unexpected: two different objects that can have the same hash (according to C++17)

Maybe you have a suggestion?

Reply:
That is the normal behavior of a hash function for not being injective https://en.wikipedia.org/wiki/Injective_function since the goal is to compress the definition domain. The hashing is not use to do the comparison, just as a heuristic to improve the data structure. At the end an equality comparison is always done.
But the real problem is that a SYCL implementations should have comparison operations working on most SYCL objects like `device`, `buffer`... and a `std::hash` specialization working.
Do we have such tests in the CTS at the first place?
Then, not using `std::sort`, `std::set` or `std::unordered_set` because of SYCL implementation bugs is futile since it is not conformant anyway. :-)
@bader My feeling is that we should implement each test the right way™ even if it does not compile today on some implementations because of some implementation limitations.

Reply:
Interesting...
How do see comparison of two sycl devices can be implemented? DPC++ implements SYCL device as a wrapper around a native handle. We could implement SYCL objects comparison operators if native handles can be compared, but I'm not sure if it's defined.
AFAIK, OpenCL device handle is just a pointer, so comparing pointer values might be unstable and we get different results sorting the same list of devices. Right?

Reply:
I have created internal ticket https://gitlab.khronos.org/sycl/Specification/-/issues/586 and we can discuss it during the next SYCL committee meeting.

Reply:
OK, from the SYCL workgroup meeting today this might be fixed in the specification.
In the meantime you can keep the O(n^2) algorithm and we can come back on this.
The problem is to remember where we will have to fix it. Adding some kind of TODO or adding a link to each test in https://gitlab.khronos.org/sycl/Specification/-/issues/586 ?

Reply:
@keryell  TODO comment was added in this commit: a3afd778782dff127cabb8f71fe4d4036f3ee5c5
Provided issue requires authorization to see it. Therefore the link on the issue was not mentioned because of not all users can see this issue.